plugins/Marketplace/LocalPackageList.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (c) 2021 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from typing import Any, Dict, Generator, TYPE_CHECKING
+from PyQt5.QtCore import pyqtSlot, Qt
+
+from UM.i18n import i18nCatalog
+
+from cura.CuraApplication import CuraApplication
+
+from .PackageList import PackageList
+from .PackageModel import PackageModel  # The contents of this list.
+
+if TYPE_CHECKING:
+    from PyQt5.QtCore import QObject
+
+catalog = i18nCatalog("cura")
+
+
+class LocalPackageList(PackageList):
+    PackageRole = Qt.UserRole + 1
+    PACKAGE_SECTION_HEADER = {
+        "installed":
+            {
+                "plugin": catalog.i18nc("@label:property", "Installed Plugins"),
+                "material": catalog.i18nc("@label:property", "Installed Materials")
+            },
+        "bundled":
+            {
+                "plugin": catalog.i18nc("@label:property", "Bundled Plugins"),
+                "material": catalog.i18nc("@label:property", "Bundled Materials")
+            }
+    }  # The section headers to be used for the different package categories
+
+    def __init__(self, parent: "QObject" = None) -> None:
+        super().__init__(parent)
+        self._manager = CuraApplication.getInstance().getPackageManager()
+        self._has_footer = False
+
+    @pyqtSlot()
+    def updatePackages(self) -> None:
+        """Update the list with local packages, these are materials or plugin, either bundled or user installed. The list
+        will also contain **to be removed** or **to be installed** packages since the user might still want to interact
+        with these.
+        """
+        self.setErrorMessage("")  # Clear any previous errors.
+        self.setIsLoading(True)
+        self._getLocalPackages()
+        self.setIsLoading(True)
+
+    def _getLocalPackages(self) -> None:
+        """ Obtain the local packages.
+
+        The list is sorted per category as in  the order of the PACKAGE_SECTION_HEADER dictionary, whereas the packages
+        for the sections are sorted alphabetically on the display name
+        """
+
+        sorted_sections = {}
+        # Filter the packages per section title and sort these alphabetically
+        for section in self._getSections():
+            packages = filter(lambda p: p.sectionTitle == section, self._allPackageInfo())
```

Comment:
I thought about this and I think you're right in the sense that it is more processing intensive and it could be optimized, but that would also reduce readability and conveying of the intent of the code. Which in my opinion outweigh the performance gain. I would be surprised if the total amount of packages on a local system is more than a 100 at any given time. I use the Python function `sorted` and `filter` which will outperform most loops and use `generator` to keep memory footprint low.

I think the performance gain of `~100 ms` will not be noticed by a user who first had to wait till the first 20 remote packages have been loaded. 

I will however have a closer look tomorrow morning  to check what the actual gain can be, because the `~100 ms` is sucked right out of my thumb.

Reply:
Looking at the pstat generated by our new CuraRemotePlugin :1st_place_medal:  shows that we're talking indeed in the magnitutde of `~10 ms` For a window which is called on occasion.
 
[cura-4.12.0-21-11-03-18_15_09.pstat.zip](https://github.com/Ultimaker/Cura/files/7469557/cura-4.12.0-21-11-03-18_15_09.pstat.zip)
![image](https://user-images.githubusercontent.com/8535734/140155170-cb34f633-988c-40b8-9057-95e1a210aac9.png)

Reply:
Won't fix expect if you feel really strong about this, but I would rather have this progress to the next column today, such that it can be picked up soon.

Reply:
I don't think that the readability is decreased if you handle it like this:

```python3
package_info = list(self._allPackageInfo())
for section in self._getSections():
    packages = filter(lambda p: p.sectionTitle == section, package_info)
    sorted_sections[section] = sorted(packages, key = lambda p: p.displayName)
```

In this case I'd even argue that using a generator might even be the "wrong" pattern in the first place; You use them when you want to save time (if there is a chance that you stop halfway through the evaluation) or when you want to save memory. 

I've profiled both codes. You're right that the change on the whole is pretty small. With the code as you wrote it it's 0.044s and with the change i proposed it's 0.003 (so 15x as fast). 

So the filter doesn't influence this at all. That also makes sense, it first generates the next package and only then checks if it needed to be filter this in the first place.

Reply:
I'm okay with applying your 15x times faster code

Reply:
Changed it into this:
```python
package_info = list(self._allPackageInfo())
sorted_sections: List[Dict[str, PackageModel]] = []
for section in self._getSections():
    packages = filter(lambda p: p.sectionTitle == section, package_info)
    sorted_sections.extend([{"package": p} for p in sorted(packages, key = lambda p: p.displayName)])
self.setItems(sorted_sections)
```
The speed increase on the function when running Yappi `LocalPackageList.updatePackages`
| original | now  |
|------------|--------|
|  14 ms  |  4 ms |

I'm curious what the new code does at your computer, since your values are maginutes higher than mine.